### PR TITLE
Adds backwards compatibility in `UsingPrecompile` contract. 

### PR DESCRIPTION
**Pull request by soloseng**
### Description

Updated `UsingPrecompile` contract to maintain backwards compatibility after L2 migration. 

Fucntions that maintained backwards compatibility query the required information from the `EpochManager` contract instead of the precompiles.

- [ ] Remove onlyL1 in validators.sol

### Other changes

Was required to add the list of elected validators to the `Epoch` struct of the EpochManager contract to keep the historical list of elected validators at a given epoch number.

### Tested

Added unit test.

### Related issues

- Fixes https://github.com/celo-org/celo-blockchain-planning/issues/642

### Backwards compatibility

for some select functions

### CODE REVIEW

**Comment by soloseng** [comment]
As of 50bec5c8a89c6e7d9026818ef1d591cf0e891b53
I was able to add the required functions to preserve backwards compatibility to the UsingRegistry contract.

Testing is not fully implemented for registry functions, but the plan was to use `EpochManager_WithMock.sol` for testing needs and if possible remove `MockEpochManager.sol`

stCELO only uses `numberValidatorsInCurrentSet` ;  `validatorSignerAddressFromCurrentSet` but I also added compatibility for the functions that perform the same query using specific blocknumber incase other contracts in the CELo ecosystem use them.

To get the epoch info based on block number, I used binary search to avoid looping through all the epoch 1x1. This was required for `getEpochNumberOfBlock()`; `validatorSignerAddressFromSet()`

We could further improve the compute cost by introducing a mapping of blocksToEpoch, 
```
mapping(uint256 => uint256) internal blockToEpoch;
``` 
but that would be a more complex change and increase the gas cost of processing an epoch as we'd have to store this info during processing.



**NOTE:** Not all functions have backwards compatibility.

File: packages/protocol/contracts/common/UsingPrecompiles.sol [hunk 1]
```diff
@@ -2,9 +2,11 @@ pragma solidity ^0.5.13;
 
 import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 import "../common/interfaces/ICeloVersionedContract.sol";
+import "../common/UsingRegistry.sol";
+import "../common/interfaces/IEpochManager.sol";
 import "../../contracts-0.8/common/IsL2Check.sol";
 
-contract UsingPrecompiles is IsL2Check {
+contract UsingPrecompiles is IsL2Check, UsingRegistry {
```
Comment (martinvol) [inline packages/protocol/contracts/common/UsingPrecompiles.sol]: UsingRegistry has a registry variable that needs to be set, if we add the dependency here then if the childs don't have UseRegistry they would break.

Comment (martinvol) [reply packages/protocol/contracts/common/UsingPrecompiles.sol]: Updating it here would also trigger redeployments of every child

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 2]
```diff
@@ -400,6 +402,38 @@ contract EpochManager is
     return epochs[epoch].lastBlock;
   }
 
+  /**
+   * @notice Returns the epoch number of a specified blockNumber.
+   * @param _blockNumber Block number of the epoch info is retreived.
+   */
+  function getEpochNumberOfBlock(uint256 _blockNumber) external view returns (uint256) {
+    (uint256 _epochNumber, , , , , ) = _getEpochByBlockNumber(_blockNumber);
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: Is there a way to do this without calling `_getEpochByBlockNumber`? If this is expected to be used in transactions rather than only off-chain view calls (by us or any third party protocol), reading the entire `elected` array into memory will waste a good bit of gas when only the epoch number is needed.

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 3]
```diff
@@ -494,4 +555,73 @@ contract EpochManager is
...
+        _elected
+      );
+    }
+
+    uint256 left = firstKnownEpoch;
+    uint256 right = currentEpochNumber - 1;
+
+    while (left <= right) {
+      uint256 mid = (left + right) / 2;
+      Epoch memory _epoch = epochs[mid];
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: Would this be much more gas efficient with only grabbing a `storage` ref (or reading only `first`/`lastBlock` into memory for the binsearch)? Not sure if the compiler optimizes away reading the other fields (including the entire `elected` array) into memory, which could incur a lot of unnecessary `SLOAD`s while searching.

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol [hunk 4]
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
...
 
   /**
-   * @notice Gets a validator address from the validator set at the given block number.
+   * @notice Gets a validator address from the current validator set.
    * @param index Index of requested validator in the validator set.
-   * @param blockNumber Block number to retrieve the validator set from.
    * @return Address of validator at the requested index.
    */
+  function validatorAddressFromCurrentSet(uint256 index) public view onlyL2 returns (address) {
+    address[] memory _elected = getEpochManager().getElected();
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/common/UsingPrecompiles.sol]: More gas optimization: add an indexed function to EpochManager that reads only the specific index, rather than reading the whole array into memory.

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol [hunk 5]
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
...
+   * @notice Gets a validator address from the validator set at the given block number.
+   * @param index Index of requested validator in the validator set.
+   * @param blockNumber Block number to retrieve the validator set from.
+   * @return Address of validator at the requested index.
+   */
+  function validatorAddressFromSet(
+    uint256 index,
+    uint256 blockNumber
+  ) public view onlyL2 returns (address) {
+    (, , , , address[] memory _elected) = getEpochManager().getEpochByBlockNumber(blockNumber);
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/common/UsingPrecompiles.sol]: Ditto

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol [hunk 6]
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
...
    * @return Size of the current elected validator set.
    */
   function numberValidatorsInCurrentSet() public view virtual returns (uint256) {
-    bytes memory out;
-    bool success;
-    (success, out) = NUMBER_VALIDATORS.staticcall(abi.encodePacked(uint256(block.number)));
-    require(success, "error calling numberValidatorsInCurrentSet precompile");
-    return getUint256FromBytes(out, 0);
+    if (isL2()) {
+      return getEpochManager().getElected().length;
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/common/UsingPrecompiles.sol]: Ditto, implementing this in EpochManager should save gas since the whole `elected` array won't be read into memory.

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol [hunk 7]
```diff
@@ -134,11 +179,17 @@ contract UsingPrecompiles is IsL2Check {
    * @return Size of the validator set.
    */
   function numberValidatorsInSet(uint256 blockNumber) public view virtual returns (uint256) {
-    bytes memory out;
-    bool success;
-    (success, out) = NUMBER_VALIDATORS.staticcall(abi.encodePacked(blockNumber));
-    require(success, "error calling numberValidatorsInSet precompile");
-    return getUint256FromBytes(out, 0);
+    if (isL2()) {
+      (, , , , address[] memory elected) = getEpochManager().getEpochByBlockNumber(blockNumber);
```
Comment (m-chrzan) [inline packages/protocol/contracts-0.8/common/UsingPrecompiles.sol]: Ditto

File: packages/protocol/contracts/common/UsingPrecompiles.sol (hunk 1 above)
Comment (m-chrzan) [reply packages/protocol/contracts/common/UsingPrecompiles.sol]: Can UsingRegistryV2 be used instead? That uses only `constants`. Or is the Registry address different on L2?

Comment (pahor167) [reply packages/protocol/contracts/common/UsingPrecompiles.sol]: Yep, unfortunately it not possible to add new inherited smart contract with storage variable to already deployed smart contract (there is very high probability that this change will break storage)

File: packages/protocol/test-sol/utils08.sol [hunk 8]
```diff
@@ -20,12 +21,20 @@ contract Utils08 {
...
+    uint256 blocksInEpoch = L2_BLOCK_IN_EPOCH;
+    blockTravel(vm, blocksInEpoch);
+    timeTravel(vm, DAY);
   }
 
   function whenL2(Vm vm) public {
     vm.etch(0x4200000000000000000000000000000000000018, abi.encodePacked(bytes1(0x01)));
   }
+
+  // This function can be also found in OpenZeppelin's library, but in a newer version than the one
```
Comment (m-chrzan) [inline packages/protocol/test-sol/utils08.sol]:
```suggestion
  // This function can be also found in OpenZeppelin's library, but in a newer version than the one we use.
```

**Review by m-chrzan — COMMENTED** [review]
Looks good, but there's a few, potentially very impactful, gas optimizations possible. There's no need to read the entire `elected` array into memory when we only need one field of the Epoch struct, or just one of the elected addresses. These should be fairly simple changes that would reduce the number of `SLOAD`s by a factor of ~100 where relevant.

**Review by martinvol — CHANGES_REQUESTED** [review]
I don't think we need to fix all precompiles in all contracts

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol [hunk 9]
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
...
   function validatorSignerAddressFromCurrentSet(
     uint256 index
   ) public view virtual returns (address) {
-    bytes memory out;
-    bool success;
-    (success, out) = GET_VALIDATOR.staticcall(abi.encodePacked(index, uint256(block.number)));
-    require(success, "error calling validatorSignerAddressFromCurrentSet precompile");
-    return address(uint160(getUint256FromBytes(out, 0)));
+    if (isL2()) {
+      return getAccounts().getValidatorSigner(validatorAddressFromCurrentSet(index));
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/common/UsingPrecompiles.sol]: this wont work if signers change mid-epoch

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol [hunk 10]
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
    * @return Epoch number.
    */
   function getEpochNumberOfBlock(uint256 blockNumber) public view returns (uint256) {
-    return epochNumberOfBlock(blockNumber, getEpochSize());
+    if (isL2()) {
+      return getEpochManager().getEpochNumberOfBlock(blockNumber);
+    } else {
+      return epochNumberOfBlock(blockNumber, getEpochSize());
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/common/UsingPrecompiles.sol]: in L2 this is seconds instead of blocks

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol (hunk 6 above)
Comment (soloseng) [reply packages/protocol/contracts-0.8/common/UsingPrecompiles.sol]: The limitation in implementing this whole function in `EpochManager`, is to avoid importing the `UsingPrecompiles` contract.

I created an `EpochManager` function that returns the number instead of getting the elected array and getting its length.

**Comment by soloseng** [comment]
Updated this PR based on our [last sync](https://github.com/celo-org/celo-blockchain-planning/issues/642#issuecomment-2397407994).

- added onlyL1 modifiers to all `UsingPrecompiles` functions. 
- created a new contract `PrecompileOverride` and it to handle the switching logic between L1 and L2. its currently being used by `Validators` contract but will need a 0.5.13 version for other core contracts still using 0.5.13 (e.g.: `Election`, Governance,EpochRewards

**Comment by soloseng** [comment]
As of commit 681a8582
Because we are wanting to keep a historical data of the elected validators account & signers, we now have to store the elected in a mapping. this causes an increase in gas when running the election, since we now have to write to 2 new storage slot (`electedAccountsOfEpoch` && `electedSignersOfEpoch` each time an election is run.

To minimize the gas cost of reading epoch info based on block number, the mapping of `electedAccountsOfEpoch` && `electedSignersOfEpoch` is kept separate than `epochs`. This allows us to retrieve the epoch info without reading the list of accounts and signers.

That being said, it would still be good to run a gas cost benchmark to know if it will be prohibitive.

usingPrecompile backwards compatibility has been added for `Election`, `Governance`, `EpochRewards` using a 0.5.13 version of the `PrecompileOverride` contract. This is because 0.5.13 does not support `virtual` && `override` function keywords.

File: packages/protocol/contracts/common/PrecompilesOverride.sol [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+ **/
+contract PrecompilesOverride is UsingPrecompiles {
+  /**
+   * @notice Returns the epoch number at a block.
+   * @param blockNumber Block number where epoch number is calculated.
+   * @return Epoch number.
+   */
+  function getEpochNumberOfBlock(uint256 blockNumber) public view returns (uint256) {
+    if (isL2()) {
+      return getEpochManager().getEpochNumberOfBlock(blockNumber);
```
Comment (pahor167) [inline packages/protocol/contracts/common/PrecompilesOverride.sol:22]: We currently dont have registry in this smart contract

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 12]
```diff
@@ -512,4 +658,86 @@ contract EpochManager is
...
+   * @notice Updates the list of elected validator signers.
+   */
+  function _setElectedSigners(address[] memory _elected) internal {
+    IAccounts accounts = getAccounts();
+    require(electedAccounts.length > 0, "Elected list length cannot be zero.");
+    if (electedSigners.length == 0) {
+      electedSigners = new address[](_elected.length);
+    }
+
+    require(_elected.length == electedSigners.length, "Account and Signer list lenght mismatch.");
```
Comment (pahor167) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: we cannot check whether _elected.length == electedSigners.length since in scenario when previously we had elected eg only 10 validators (because there were only 10 validators registered) and during this epoch new validators registered then it is valid to have different count of electedSigners from before

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 3 above)
Comment (soloseng) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: This struct no longer includes the election history

File: packages/protocol/contracts/common/UsingPrecompiles.sol (hunk 1 above)
Comment (soloseng) [reply packages/protocol/contracts/common/UsingPrecompiles.sol]: using `PrecompileOverride` contract, as discussed [here](https://github.com/celo-org/celo-blockchain-planning/issues/642#issuecomment-2397407994)

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 13]
```diff
@@ -506,4 +649,82 @@ contract EpochManager is
...
+        );
+      } else if (_blockNumber < _epochFirstBlock) {
+        right = mid - 1;
+      } else {
+        left = mid + 1;
+      }
+    }
+
+    revert("No matching epoch found for the given block number.");
+  }
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:724]: This function is very likely to run out of gas, is it (and the functions that use it) really needed?

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 12 above)
Comment (martinvol) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: I saw this conversation marked as resolved, I think the issue still remains

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 14]
```diff
@@ -506,4 +649,82 @@ contract EpochManager is
...
+
+  /**
+   * @notice Returns the epoch info of a specified blockNumber.
+   * @param _blockNumber Block number of the epoch info is retreived.
+   * @return firstEpoch The first block of the given block number.
+   * @return lastBlock The first block of the given block number.
+   * @return startTimestamp The starting timestamp of the given block number.
+   * @return rewardsBlock The reward block of the given block number.
+   */
+  function _getEpochByBlockNumber(
```
Comment (martinvol) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:626]: This function is very likely to run out of gas, is it (and the functions that use it) really needed?

I also think that it has quite some edge cases, would be difficult to test them all.

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 15]
```diff
@@ -51,13 +51,18 @@ contract EpochManager is
...
+  address[] internal electedAccounts;
+  address[] internal electedSigners;
   address public oracleAddress;
-  address[] public elected;
 
   mapping(address => uint256) public processedGroups;
 
   EpochProcessState public epochProcessing;
   mapping(uint256 => Epoch) internal epochs;
+  mapping(uint256 => address[]) internal electedAccountsOfEpoch;
```
Comment (pahor167) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: This variable doesnt seem to be set anywhere now

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 16]
```diff
@@ -51,13 +51,18 @@ contract EpochManager is
...
+  // Electeds in the L1 assumed signers can not change during the epoch
+  // so we keep a copy
+  mapping(uint256 => address[]) internal electedSignersOfEpoch;
```
Comment (pahor167) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: Same here

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 17]
```diff
@@ -51,13 +51,18 @@ contract EpochManager is
 
   uint256 public firstKnownEpoch;
   uint256 internal currentEpochNumber;
+  address[] internal electedAccounts;
```
Comment (pahor167) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: Since this contract is already deployed on alfajores we should try not to break storage

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 18]
```diff
@@ -240,7 +251,7 @@ contract EpochManager is
         );
         processedGroups[group] = epochRewards == 0 ? type(uint256).max : epochRewards;
       }
-      delete elected[i];
```
Comment (pahor167) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:243]: We should be deleting also signers here

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 19]
```diff
@@ -264,8 +275,17 @@ contract EpochManager is
...
-    _epochProcessing.status = EpochProcessStatus.NotStarted;
+
+    address[] memory _newlyElected = election.electValidatorAccounts();
+
+    electedAccounts = _newlyElected;
+
+    _setElectedSigners(_newlyElected);
+
+    EpochProcessState memory _epochProcessingEmpty;
+    epochProcessing = _epochProcessingEmpty;
```
Comment (pahor167) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:284]: Do we need to zero this out here ?

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 17 above)
Comment (soloseng) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: made public

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 19 above)
Comment (soloseng) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:284]: This was a suggestion from @martinvol. I don't have any issues with zeroing it out at the end of the epoch process.

**Review by martinvol — APPROVED** [review]

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 14 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:626]: comment added

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 19 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:284]: leaving as it is now
